compiler/src/dotty/tools/dotc/core/StagingContext.scala

```diff
@@ -18,8 +24,27 @@ object StagingContext {
   def quoteContext(implicit ctx: Context): Context =
     ctx.fresh.setProperty(QuotationLevel, level + 1)
 
+  /** Context with an incremented quotation level and pushes a refecence to a QuoteContext on the quote context stack */
+  def pushQuoteContext(qctxRef: tpd.Tree)(implicit ctx: Context): Context =
+    val old = ctx.property(QuoteQontextStack).getOrElse(List.empty)
+    ctx.fresh.setProperty(QuotationLevel, level + 1)
+             .setProperty(QuoteQontextStack, qctxRef :: old)
+
   /** Context with a decremented quotation level. */
   def spliceContext(implicit ctx: Context): Context =
     ctx.fresh.setProperty(QuotationLevel, level - 1)
-}
 
+  /** Context with a decremented quotation level and pops the Some of top of the quote context stack or None if the stack is empty.
+   *  The quotation stack could be empty if we are in a top level splice or an eroneous splice directly witin a top level splice.
+   */
+  def popQuoteContext()(implicit ctx: Context): (Option[tpd.Tree], Context) =
+    val ctx1 = ctx.fresh.setProperty(QuotationLevel, level - 1)
```

Comment:
The property `QuotationLevel` is also mutated in `spliceContext` and `quoteContext`. This may cause accidental errors when used without care. Maybe remove  `spliceContext` and `quoteContext`, since they are not used.

Reply:
We use those in `PCPCheckAndHeal` and `ReifyQuotes`. There we only need to track the level.